src/linear_algebra/eigenspace.lean

```diff
@@ -0,0 +1,238 @@
+/-
+Copyright (c) 2020 Alexander Bentkamp. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alexander Bentkamp.
+-/
+
+import field_theory.algebraic_closure
+import linear_algebra.finsupp
+
+/-!
+# Eigenvectors and eigenvalues
+
+This file defines eigenspaces and eigenvalues.
+
+An eigenspace of a linear map `f` for a scalar `μ` is the kernel of the map `(f - μ • id)`.
+The elements of an eigenspace are eigenvectors `x`. They have the property `f x = μ • x`.
+If there are nonzero eigenvectors, the scalar `μ` is called an eigenvalue.
```

Comment:
The standard mathematical usage of the word "eigenvector" is that it is by definition non-zero. If we want to change it here then that's...kind-of fine, I guess...but I think that it should be explicitly flagged in big red flashing letters that we are not using the standard terminology. The eigenspace for lambda is the disjoint union of 0 and the eigenvectors for lambda :-/ We might not like it, but that is the meaning of these words :-/

Reply:
Hehe, on the previous version of this pull request, I got the opposite feedback (#3864) :D

Just for the record: It might be in the minority, but this paper defines eigenvectors as possibly being 0: https://www.maa.org/sites/default/files/pdf/awards/Axler-Ford-1996.pdf
Without a flashing warning sign.

Note that currently, I don't even formally define eigenvectors, but only eigenspaces and eigenvalues. The sentence in the documentation could simply be removed. So we can go either way with the definition of `eigenvector` (if we even need one).

I don't think we need a definition saying `eigenvector f μ x := x ∈ eigenspace f μ`. Writing `x ∈ eigenspace f μ` is just as short. So instead, we could define `eigenvector f μ x := x ≠ 0 ∧ x ∈ eigenspace f μ`, following the mathematical majority. This definition would also shorten the statement of the lemma `eigenvectors_linear_independent`.

@jcommelin @sgouezel What do you think?

Reply:
People are definitely going to be looking for the definition of eigenvector in this file, and it's a short definition, so I would like to see `def is_eigenvector := x ≠ 0 ∧ x ∈ eigenspace f μ` right after `eigenspace`. I'd also like to point out that the only usage of `x ∈ eigenspace` in this PR (`eigenvectors_linear_independent`) also has the hypothesis `x ≠ 0` for those `x`, so it looks like a natural concept even in this file.

Reply:
Sounds good. I did that.

Reply:
To me, an eigenvector can definitely be zero. At least, that's what I learnt in school. Maybe that's a French thing, but opening a random reference book (Lang, Algebra), I can read on Page 562 "By an eigenvector $$w$$ of $$A$$ in $$E$$, one means an element $$w \in E$$ such that there exists an element $$\lambda \in k$$ for which $$A w = \lambda w$$. Same thing in Bourbaki (Algebra, Chapter 7). So, the only conclusion I can draw is that there is nothing universal on the definition, and in any case we should put a warning saying which definition we use.

My preference goes to allowing `0` as an eigenvector, just because it makes an easier to use definition (the sum of eigenvectors for some eigenvalue is an eigenvector, and so on).

Reply:
I kept the definition that @Vierkantor proposed, but added a warning and a note that one can write `x ∈ eigenspace f μ` to get eigenvectors including `0`.